File: src/models/dto/records.rs
use std::borrow::Cow;

use chrono::NaiveDate;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct NewProductionRecord<'a> {
    pub tag_id: Cow<'a, str>,
    pub production_type: Cow<'a, str>,
    pub quantity: i64,
    pub unit: Cow<'a, str>,
    pub quality: Option<Cow<'a, str>>,
    pub notes: Option<Cow<'a, str>>,
    pub record_date: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct UpdateProductionRecord<'a> {
    pub production_type: Option<Cow<'a, str>>,
    pub quantity: Option<i64>,
    pub unit: Option<Cow<'a, str>>,
    pub quality: Option<Cow<'a, str>>,
    pub notes: Option<Cow<'a, str>>,
    pub record_date: Option<NaiveDate>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct NewHealthRecord<'a> {
    pub tag_id: Cow<'a, str>,
    pub record_date: Cow<'a, str>,
    pub condition: Cow<'a, str>,
    pub description: Cow<'a, str>,
    pub treatment: Cow<'a, str>,
    pub severity: Cow<'a, str>,
    pub status: Cow<'a, str>,
    pub medicine: Option<Cow<'a, str>>,
    pub dosage: Option<Cow<'a, str>>,
    pub cost: Option<i64>,
    pub performed_by: Option<Cow<'a, str>>,
    pub prognosis: Option<Cow<'a, str>>,
    pub notes: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct UpdateHealthRecord<'a> {
    pub record_date: Option<NaiveDate>,
    pub condition: Option<Cow<'a, str>>,
    pub description: Option<Cow<'a, str>>,
    pub treatment: Option<Cow<'a, str>>,
    pub severity: Option<Cow<'a, str>>,
    pub status: Option<Cow<'a, str>>,
    pub medicine: Option<Cow<'a, str>>,
    pub dosage: Option<Cow<'a, str>>,
    pub cost: Option<i64>,
    pub performed_by: Option<Cow<'a, str>>,
    pub prognosis: Option<Cow<'a, str>>,
    pub notes: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct NewWeightRecord<'a> {
    pub tag_id: Cow<'a, str>,
    pub record_date: NaiveDate,
    pub mass: i64,
    pub unit: Cow<'a, str>,
    pub status: Cow<'a, str>,
    pub notes: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct UpdateWeightRecord<'a> {
    pub record_date: Option<NaiveDate>,
    pub mass: Option<i64>,
    pub previous_mass: Option<i64>,
    pub unit: Option<Cow<'a, str>>,
    pub status: Option<Cow<'a, str>>,
    pub notes: Option<Cow<'a, str>>,
}

File: frontend/src/lib/schemas/records.ts
import { z } from "zod";

export const newProductRecordSchema = z.object({
	tagId: z.string().min(2).max(10),
	productionType: z.string().min(3).max(50),
	quantity: z.coerce.number(),
	unit: z.string().min(1, { message: "unit is required" }).max(10),
	quality: z.string().optional().nullable(),
	notes: z.string().optional().nullable(),
	recordDate: z.date().optional().nullable(),
});

export type NewProductRecord = z.infer<typeof newProductRecordSchema>;

export const newHealthRecordSchema = z.object({
	recordDate: z.date(),
	condition: z.enum(["injury", "vaccination", "fever", "checkup", "infection"]),
	status: z.enum(["recovered", "recovering", "active", "deceased", "worsened"]),
	severity: z.enum(["high", "medium", "low"]),
	notes: z.string().optional().nullable(),
	treatment: z.string(),
	dosage: z.string().optional(),
	medicine: z.string().optional(),
	tagId: z.string().nullable(),
	description: z.string(),
	performedBy: z.string().optional(),
	cost: z.coerce.number(),
});

export type NewHealthRecord = z.infer<typeof newHealthRecordSchema>;

export const newWeightRecordSchema = z.object({
	tagId: z.string().min(2, { message: "Tag ID requires 2 characters" }),
	recordDate: z.date(),
	mass: z.coerce.number(),
	unit: z.enum(["kg", "lb"]),
	status: z.enum(["overweight", "underweight", "normal"]),
	notes: z.string().optional().nullable(),
});

export type NewWeightRecord = z.infer<typeof newWeightRecordSchema>;

export const newOffspringRecordSchema = z.object({
	name: z
		.string()
		.min(1, "Offspring name is required")
		.max(50, "Name must be less than 50 characters"),

	tagId: z
		.string()
		.optional()
		.or(z.literal(""))
		.transform((val) => (val === "" ? undefined : val)),

	gender: z.enum(["male", "female"], {
		required_error: "Please select a gender",
	}),

	birthDate: z
		.string()
		.or(z.date())
		.transform((val) => {
			if (typeof val === "string") {
				return new Date(val);
			}
			return val;
		})
		.refine((date) => date <= new Date(), {
			message: "Birth date cannot be in the future",
		}),

	breed: z
		.string()
		.optional()
		.or(z.literal(""))
		.transform((val) => (val === "" ? undefined : val)),

	color: z
		.string()
		.optional()
		.or(z.literal(""))
		.transform((val) => (val === "" ? undefined : val)),

	weight: z
		.number()
		.positive("Weight must be a positive number")
		.max(1000, "Weight seems unrealistic")
		.optional()
		.or(
			z.string().transform((val) => {
				const num = Number.parseFloat(val);
				return Number.isNaN(num) ? undefined : num;
			}),
		),

	healthStatus: z
		.enum(["healthy", "sick", "recovering", "deceased"], {
			required_error: "Please select a health status",
		})
		.default("healthy"),

	notes: z
		.string()
		.max(1000, "Notes must be less than 1000 characters")
		.optional()
		.or(z.literal(""))
		.transform((val) => (val === "" ? undefined : val)),

	parentId: z.string().optional(), // This will be added by the component
});

export type NewOffspringRecord = z.infer<typeof newOffspringRecordSchema>;
